### PR TITLE
Laboratory work #3, Milana Hodzhametova - 21FPL2

### PR DESCRIPTION
**Pull request by miilaanaa**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Nothing meaningful to review yet :<

**Review by marina-kaz — COMMENTED** [review]
No changes to check :<

**Review by marina-kaz — COMMENTED** [review]
Nothing new to review today :<

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -152,7 +172,16 @@ def decode(self, encoded_tokens: tuple[int, ...]) -> Optional[tuple[str, ...]]:
                 sequence of string tokens
         In case of out-of-dictionary input data, None is returned
         """
-        pass
+        if not encoded_tokens:
+            return None
+        for token in encoded_tokens:
+            if token not in self._id2word.keys():
+                return None
+        tokens_list = [self._id2word[key] for token in encoded_tokens for key in self._id2word if token == key]
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:180]: try to find a better way to do this, this is awkward and redundant

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -215,7 +257,9 @@ def __init__(self) -> None:
         """
         Constructs all the necessary attributes for the adjacency matrix graph object
         """
-        pass
+        self._matrix = [[]]
+        self._positions = {}
+        self._position_weights = {}
 
     # Step 4.2
     def add_edge(self, vertex1: int, vertex2: int) -> int:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:265]: Matrix is only supposed to contain zeros and ones, think of a matrix in math
Do not append vertices themselves in the matrix

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -385,7 +472,14 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vertex in [vertex1, vertex2]:
+            if vertex not in self._edges:
+                self._edges[vertex] = []
+        self._edges[vertex1].append(vertex2)
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:480]: what if these vertices are already connected?

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -530,19 +642,7 @@ def train(self) -> None:
...
-        for _ in range(0, self._max_iter):
-            prev_score = self._scores.copy()
-            for scored_vertex in vertices:
-                incidental_vertices = [vertex for vertex in vertices
-                                       if self._graph.is_incidental(scored_vertex, vertex) == 1]
-                self.update_vertex_score(scored_vertex, incidental_vertices, prev_score)
-            abs_score_diff = [abs(i - j) for i, j in zip(prev_score.values(), self._scores.values())]
-            if sum(abs_score_diff) <= self._convergence_threshold:
-                break
+        pass
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:645]: train is written for you, you can use it

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -564,7 +664,7 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
-        pass
+        return tuple(sorted(self._scores.keys(), key=lambda key: self._scores[key], reverse=True)[:n_keywords])
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:667]: what if 2 or more vertices have the same score?

**Review by marina-kaz — COMMENTED** [review]
Great job!
Keep going!!

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -356,18 +418,22 @@ def __init__(self) -> None:
...
         """
-        pass
+        self._edges = {}
+        self._positions = {}
+        self._position_weights = {}
+
 
     # Step 7.2
-    def get_vertices(self) -> tuple[int, ...]:
+    def get_vertices(self) -> tuple[str, ...]:
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:427]: why did you have to modify the typing annotation? 
it must stay the same as in the task description as it is a part of the task...

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -564,7 +657,8 @@ def get_top_keywords(self, n_keywords: int) -> tuple[int, ...]:
             tuple[int, ...]
                 top n most important tokens in the encoded text
         """
-        pass
+        sorted_key = sorted(self._scores)
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:660]: although dictionaries are in fact ordered in the later versions of python, it is better to establish secondary sorting order using `key` argument of the `sorted` fuction

File: lab_3_keywords_textrank/main.py [hunk 8]
```diff
@@ -443,15 +527,19 @@ def fill_positions(self, tokens: tuple[int, ...]) -> None:
             tokens : tuple[int, ...]
                 sequence of tokens
         """
-        pass
+        for index, word in enumerate(tokens):
+            self._positions[word] = self._positions.get(word, []) + [index + 1]
```
Comment (marina-kaz) [inline lab_3_keywords_textrank/main.py:531]: are you sure it works as you expect?

**Review by marina-kaz — COMMENTED** [review]
Great job!
A bit to improve

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (miilaanaa) [reply lab_3_keywords_textrank/main.py:427]: asked to change because the return was highlighted

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**